ingeniamotion/fsoe.py

```diff
@@ -0,0 +1,266 @@
+
+    """
+
+    DEFAULT_WATCHDOG_TIMEOUT_S = 1
+    DEFAULT_FSOE_SLAVE_ADDRESS = 0
+
+    def __init__(self, motion_controller: "MotionController") -> None:
+        self.logger = ingenialogger.get_logger(__name__)
+        self.__mc = motion_controller
+        self.__handlers: Dict[str, FSoEMasterHandler] = {}
+        self.__next_connection_id = 1
+
+    def create_fsoe_master_handler(
+        self,
+        servo: str = DEFAULT_SERVO,
+        fsoe_master_watchdog_timeout: float = DEFAULT_WATCHDOG_TIMEOUT_S,
+    ) -> None:
+        """Create an FSoE Master handler linked to a Safe servo drive.
+
+        Args:
+            servo: servo alias to reference it. ``default`` by default.
+            fsoe_master_watchdog_timeout: The FSoE master watchdog timeout in seconds.
+
+        """
+        # TODO: use function to read the FSoE slave address (INGM-446)
+        slave_address = self.DEFAULT_FSOE_SLAVE_ADDRESS
+        master_handler = FSoEMasterHandler(
+            slave_address, self.__next_connection_id, fsoe_master_watchdog_timeout
+        )
+        self.__handlers[servo] = master_handler
+        self.__next_connection_id += 1
+
+    def start_master(self, start_pdos: bool = False) -> None:
+        """Start all the FSoE Master handlers.
+
+        Args:
+            start_pdos: if ``True``, start the PDO exchange, if ``False``
+                the PDO exchange should be started after. ``False`` by default.
+
+        """
+        if start_pdos:
+            self.set_pdo_maps_to_slaves()
+            self.subscribe_to_pdo_thread_events()
+            self.__mc.capture.pdo.start_pdos()
+        for servo, master_handler in self.__handlers.items():
+            master_handler.start()
+
+    def subscribe_to_pdo_thread_events(self) -> None:
+        """Subscribe to the PDO thread events.
+
+        This allows to send the Safety Master PDU and to retrieve the Safety Slave PDU.
+
+        """
+        self.__mc.capture.pdo.subscribe_to_send_process_data(self._get_request)
+        self.__mc.capture.pdo.subscribe_to_receive_process_data(self._set_reply)
+        self.__mc.capture.pdo.subscribe_to_exceptions(self._pdo_thread_exception_handler)
+
+    def unsubscribe_from_pdo_thread_events(self) -> None:
+        """Unsubscribe from the PDO thread events."""
+        self.__mc.capture.pdo.unsubscribe_to_send_process_data(self._get_request)
+        self.__mc.capture.pdo.unsubscribe_to_receive_process_data(self._set_reply)
+        self.__mc.capture.pdo.unsubscribe_to_exceptions(self._pdo_thread_exception_handler)
+
+    def set_pdo_maps_to_slaves(self) -> None:
+        """Set the PDOMaps to be used by the Safety PDUs to the slaves."""
+        for servo, master_handler in self.__handlers.items():
+            rpdo_map = master_handler.safety_master_pdu_map
+            tpdo_map = master_handler.safety_slave_pdu_map
+            self.__mc.capture.pdo.set_pdo_maps_to_slave(rpdo_map, tpdo_map, servo)
```

Comment:
I'm not sure we will have (or be able) to do this. The map of the slave I think is not configurable

Reply:
The PDOs will already be mapped on the slave, by calling this method the PDO will be assigned.